File: udprx.py
import argparse
import socket
import time
import struct


def go_realtime(sched_fifo_priority):
    """
    struct sched_param p = {};
    p.sched_priority = sched_fifo_priority;

    int rc = sched_setscheduler(0, SCHED_FIFO, &p);

    if (rc){
            perror("failed to set RT priority");
    }
    """
    # TODO
    pass


class Timer:
    def __init__(self):
        # Store the initial time in nanoseconds
        self.t1 = time.perf_counter_ns()

    def now(self):
        self.t1 = time.perf_counter_ns()

    def timeus(self):
        # Get the current time in nanoseconds
        t2 = time.perf_counter_ns()
        
        # Calculate the difference and convert to microseconds
        # 1 microsecond = 1000 nanoseconds
        return (t2 - self.t1) // 1000

def elapsed_time(t0, t1):
    return t1 - t0

if __name__ == "__main__":
    UDP_IP = "127.0.0.1"
    UDP_PORT = 53676
    PACKET_SIZE = 9000
    SOCKET_BUFFER_SIZE = 2000000
    SAMPLES_PER_PACKET = 1
    SAMPLE_SIZE_IN_BYTES = 64
    MAX_ERRORS = 9

    parser = argparse.ArgumentParser(
        prog="udprx", description="Receives UDP packets", epilog="UDP packet receiver"
    )

    parser.add_argument(
        "-p",
        "--port",
        default=UDP_PORT,
        type=int,
        help=f"UDP receive port (default: {UDP_PORT})",
    )
    parser.add_argument(
        "-s",
        "--size",
        default=PACKET_SIZE,
        type=int,
        help=f"User data size (default: {PACKET_SIZE})",
    )
    parser.add_argument(
        "-b,",
        "--socket_buffer_size",
        default=SOCKET_BUFFER_SIZE,
        type=int,
        help=f"socket buffer size in bytes (default: {SOCKET_BUFFER_SIZE})",
    )
    parser.add_argument("--spp", default=SAMPLES_PER_PACKET, type=int, help=f"Samples per packet (default: {SAMPLES_PER_PACKET})")
    parser.add_argument("--ssb", default=SAMPLE_SIZE_IN_BYTES, type=int, help=f"Sample size in bytes (default: {SAMPLE_SIZE_IN_BYTES})")
    parser.add_argument(
        "-c",
        "--count_column",
        default=-1,
        type=int,
        help="Count column (indexed from 0)",
    )
    parser.add_argument(
        "-t",
        "--step",
        default=1,
        type=int,
        help="Count step (default:1), but may be decimated",
    )
    parser.add_argument(
        "-R", "--rt_prio", default=0, type=int, help="set POSIX RT priority (0: no set)"
    )
    parser.add_argument(
        "-o", "--output", default=-1, type=int, help="1: output data to stdout"
    )
    parser.add_argument("-q", "--quiet", default=0, type=int, help="1: stop reporting")
    parser.add_argument(
        "-S",
        "--max_samples",
        default=0,
        type=int,
        help="stop after this many samples, 0: no limit",
    )
    parser.add_argument(
        "-M",
        "--max_errs",
        default=MAX_ERRORS,
        type=int,
        help=f"stop after this many errors (default: {MAX_ERRORS})",
    )
    parser.add_argument(
        "-a",
        "--local_address",
        default="0.0.0.0",
        help="optional local address\ne.g. multiple NICs, one port",
    )
    parser.add_argument(
        "-v", "--verbose", default=0, type=int, help="increase to get more chatty"
    )

    args = parser.parse_args()
    print(args)

    data_size = args.ssb * args.spp
    rx_bytes = 0
    rx_packets = 0
    rx_bytes_total = 0
    deviation = False
    error_count = 0
    rx_packets_last_error = 0
    packets_lost = 0
    interval_us = 10000000
    interval_s = interval_us / 1000000
    B1M = 1000000

    update_timer = Timer()
    usecs = update_timer.timeus()

    # print(f"Expecting data size: {data_size}")

    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.bind((args.local_address, args.port))

    if args.rt_prio:
        go_realtime(args.rt_prio)

    # TODO: figure out how to make this run forever is max_samples isn't set
    samples = 0
    while True:
        # Condition to stop the loop
        if args.max_samples != 0 and samples >= args.max_samples:
            print(f"Reached max samples ({samples}). Stopping.")
            break

        data, addr = sock.recvfrom(args.socket_buffer_size)

        read_size = len(data)

        rx_bytes += read_size
        rx_packets += 1
        if rx_packets == 1:
            tick = time.time()

        if args.count_column >= 0:
            # Grab the first SPAD Count in case we're not starting from 1
            if rx_packets == 1:
                # print("rx_packets == 1")
                offset = 4 * args.count_column
                spad_tracker = struct.unpack_from("<I", data, offset)[0]
                # print(f"spad_tracker = {spad_tracker}")

            for i in range(0, args.spp):
                # print(f"i={i}")
                spad_index = i * args.ssb + 4 * args.count_column
                # print(f"spad_index = {spad_index} and len(data)={len(data)}")

                spad_count = struct.unpack_from("<I", data, spad_index)[0]
                # print(f"spad_index={spad_index}")
                # print(f"spad_count = {spad_count}")
                if (args.verbose and (samples + i < args.spp+1)) or deviation == True:
                    if deviation:
                        dev_or_ini = "dev"
                    else:
                        dev_or_ini = "ini"
                    print(f"{spad_count:010x} {spad_count} {dev_or_ini}")
                    deviation = False
                if (spad_tracker) != spad_count:
                    # print(f"spad_tracker=={spad_tracker} and spad_count={spad_count}")
                    deviation = True
                    error_count = error_count + 1
                    print(
                        f"Deviation! Err={error_count} Expected={spad_tracker} Received={spad_count} "
                        f"Packets={rx_packets-1} PacketsSinceLastError={rx_packets - 1 - rx_packets_last_error} "
                        f"SampleJump={spad_count - spad_tracker} PacketsLost={(spad_count - spad_tracker) // args.spp} BytesLost={args.ssb * (spad_count - spad_tracker)}\n",
                    )
                    packets_lost = packets_lost + (spad_count - spad_tracker) / args.spp
                    rx_packets_last_error = rx_packets
                    spad_tracker = spad_count  # Ignore error, reinitialise tracker variable
                    if error_count > args.max_errs:
                        print("Maximum error count reached, quitting\n")
                        exit(0)
                spad_tracker = spad_tracker + args.step
        samples += args.spp

        if args.output >= 0:
            raise NotImplementedError(
                "Outputting to stdout from this program is not yet implemented"
            )

        if (rx_packets % 100) == 0:
            usecs = update_timer.timeus()

        print(f"usecs={usecs} interval_us={interval_us}")
        if args.verbose and usecs >= interval_us:
            rx_bytes_total += rx_bytes
            tock = time.time()
            elapsed_str = elapsed_time(tick,tock)
            if (args.count_column >= 0):
                print(f"RxRate: % {rx_bytes / B1M / interval_s}MB/s (total: % {rx_bytes_total / B1M}  MB) Elapsed {elapsed_str} PktRec = {rx_packets} ErrCount = {error_count} PktsLost = {packets_lost} PER {(1.0 * packets_lost / (rx_packets + packets_lost))} \n")
            else:
                print(f"Rx rate: {rx_bytes * 8.0 / (usecs / 1000000.0) / B1M} Mbps, rx % {rx_bytes / B1M / interval_s} MB/s (total: % {rx_bytes_total / B1M} MB), Elapsed {elapsed_str} , PktRec = {rx_packets} \n")
            rx_bytes = 0
            update_timer.now()
            usecs = update_timer.timeus()

File: udprx.cpp
#define __STDC_FORMAT_MACROS 1
#include <iostream>
#include <string>
#include <vector>
#include <CLI/CLI.hpp>
#include <common/Socket.h>
#include <common/Timer.h>
#include "rt.h"

#ifdef _WIN32
  #include <winsock2.h>
  #pragma comment(lib, "ws2_32.lib") // Link with the Winsock library
#else
  #include <stdio.h>
  #include <unistd.h>
#endif

// Get current date/time, format is YYYY-MM-DD.HH:mm:ss
const std::string currentDateTime() {
    time_t     now = time(0);
    struct tm  tstruct;
    char       buf[80];
    tstruct = *localtime(&now);
    // Visit http://en.cppreference.com/w/cpp/chrono/c/strftime
    // for more information about date/time format
    strftime(buf, sizeof(buf), "%Y-%m-%d.%X", &tstruct);

    return buf;
}


struct {
  int UDPPort{9000};
  int DataSize{9000};
  int SocketBufferSize{2000000};
  int SamplesPerPacket{1};
  int SampleSizeBytes{64};
  int CountColumn{-1};
  int CountStep{1};
  int RtPrio{0};
  int outfd{-1};
  int quiet{0};
  int maxerrs{9};
  uint64_t maxsamples{0};
  std::string local_address{"0.0.0.0"};
  int verbose{0};
} Settings;
  
void fmtElapsedTime(char *str, int tick, int tock) {
  int elapsed = tock - tick;
  int hours = elapsed / 3600;
  
  elapsed %= 3600;
  int minutes = elapsed / 60 ;
 
  elapsed %= 60;
  int seconds = elapsed;
  
  snprintf(str,10,"%02i:%02i:%02i",hours,minutes,seconds);  
}

CLI::App app{"UDP receiver with 32 bit sequence number check."};

int main(int argc, char *argv[]) {

    #ifdef _WIN32
      // Must initialize winsock
      WSADATA wsaData;
      int iResult = WSAStartup(MAKEWORD(2,2), &wsaData);
      if (iResult != 0) {
        printf("WSAStartup failed: %d\n", iResult);
        return 1;
      }
    #endif

    app.add_option("-p, --port", Settings.UDPPort, "UDP receive port")->capture_default_str();
    //app.add_option("-s, --size", Settings.DataSize, "User data size")->capture_default_str();
    app.add_option("-b, --socket_buffer_size", Settings.SocketBufferSize, "socket buffer size (bytes)")->capture_default_str();
    app.add_option("--spp", Settings.SamplesPerPacket, "Samples per packet")->capture_default_str();
    app.add_option("--ssb", Settings.SampleSizeBytes, "Sample size (bytes)")->capture_default_str();
    app.add_option("-c, --count_column", Settings.CountColumn, "Count column (indexed from 0)")->capture_default_str();
    app.add_option("-t, --step", Settings.CountStep, "Count step (default:1), but may be decimated")->capture_default_str();
    app.add_option("-R, --rt_prio", Settings.RtPrio, "set POSIX RT priority (0: no set)")->capture_default_str();
    app.add_option("-o, --output", Settings.outfd, "1: output data to stdout")->capture_default_str();
    app.add_option("-q, --quiet", Settings.quiet, "1: stop reporting")->capture_default_str();
    app.add_option("-S, --max_samples", Settings.maxsamples, "stop after this many samples, 0: no limit")->capture_default_str();
    app.add_option("-M, --max_errs", Settings.maxerrs, "stop after this many errors")->capture_default_str();
    app.add_option("-a, --local_address", Settings.local_address, "optional local address\neg multiple NICs, one port")->capture_default_str();
    app.add_option("-v, --verbose", Settings.verbose, "increase to get more chatty")->capture_default_str();    CLI11_PARSE(app, argc, argv);

    static const int BUFFERSIZE{9200};
    char buffer[BUFFERSIZE];
    uint64_t RxBytesTotal{0};
    uint64_t RxBytes{0};
    uint64_t RxPackets{0};
    uint64_t RxPacketsLastError{0};
    uint32_t SpadTracker{1};
    uint32_t SpadCount{1};
    uint32_t PktsLost{0};
    int ErrCount{0};
    int SpadIndex{0};
    char elapsed_str[10];
    const int intervalUs = 10000000;
    const int interval_s = intervalUs / 1000000;
    const int B1M = 1000000;
    time_t tick{0};
    time_t tock{0};
    bool deviation = false;

    if (Settings.RtPrio){
        goRealTime(Settings.RtPrio);
    }
    Socket::Endpoint local(Settings.local_address, Settings.UDPPort);
    UDPReceiver Receive(local);
    Receive.setBufferSizes(Settings.SocketBufferSize, Settings.SocketBufferSize);
    Receive.setRecvTimeout(30, 0); // set a 30 second timeout on socket
    Receive.printBufferSizes();

    Settings.DataSize = Settings.SampleSizeBytes*Settings.SamplesPerPacket;

    Timer UpdateTimer;
    auto USecs = UpdateTimer.timeus();

    // Print the config as a string (can also be used to save the config to a file)
    std::cout << "--- Configuration from arguments ---" << std::endl;
    std::string config_string = app.config_to_str();
    std::cout << config_string << std::endl;

    // Print the config to a string with any default arguments included
    std::cout << "--- Configuration from defaults ---" << std::endl;
    std::string full_config_string = app.config_to_str(true, true);
    std::cout << full_config_string << std::endl;


    for (uint64_t samples = 0;
         Settings.maxsamples == 0 || samples < Settings.maxsamples;
         samples+=Settings.SamplesPerPacket)
    {
        int ReadSize = Receive.receive(buffer, BUFFERSIZE);
        if (ReadSize == -1) {
            printf("exiting as no data received...");
            exit(1);
        }

        assert(ReadSize > 0);
        assert(ReadSize == Settings.DataSize);

        RxBytes += ReadSize;
        RxPackets++;
        if (RxPackets == 1){
            tick = time(0);
            printf("tick");
        }

        if (Settings.CountColumn >= 0){
            // Grab the first SPAD Count in case we're not starting from 1
            if (RxPackets == 1) {
                SpadTracker = *((uint32_t *)( buffer + 4*Settings.CountColumn ));
            }

            for (int i = 0; i <= Settings.SamplesPerPacket-1; i++) {
                SpadIndex = i*Settings.SampleSizeBytes + 4*Settings.CountColumn;

                SpadCount = *((uint32_t *)( buffer + SpadIndex ));
                if ((Settings.verbose && samples+i < 5) || deviation == true) {
                    fprintf(stderr, "%#010x    %i %s\n", SpadCount, SpadCount, deviation? "dev":"ini");
                deviation = false;
                }
                if (SpadTracker != SpadCount) {
                    deviation = true;
                    ErrCount = ErrCount + 1;
                    fprintf(stderr, "Deviation! Err=%i Expected=%i Received=%i " \
                            "Packets=%li PacketsSinceLast=%li " \
                            "SampleJump=%i PacketsLost=%i Bytes=%i\n", 
                            ErrCount,
                            SpadTracker,
                            SpadCount,
                            RxPackets-1,
                            RxPackets-1-RxPacketsLastError,
                            (SpadCount - SpadTracker),
                            (SpadCount - SpadTracker)/Settings.SamplesPerPacket, 
                            (Settings.SampleSizeBytes * (SpadCount - SpadTracker)));
                    fflush(stderr);
                    PktsLost = PktsLost + (SpadCount - SpadTracker)/Settings.SamplesPerPacket;
                    RxPacketsLastError = RxPackets;
                    SpadTracker = SpadCount; // Ignore error, reinitialise tracker variable
                    if (ErrCount > Settings.maxerrs) {
                        fprintf(stderr, "Maximum error count reached, quitting\n");
                        exit(0);
                    }
                }
                SpadTracker = SpadTracker + Settings.CountStep;
            }
        }
        if (Settings.outfd >= 0){
            #ifdef _WIN32
              // TODO: implement on windows
              printf("Not implemented on windows version");
            #else
              write(Settings.outfd, buffer, ReadSize);
            #endif
        }
        if ((RxPackets % 100) == 0){
            USecs = UpdateTimer.timeus();
        }

        if (!Settings.quiet && USecs >= intervalUs) {
            RxBytesTotal += RxBytes;
            time_t tock = time(0);
            fmtElapsedTime(elapsed_str,tick,tock);
            fprintf(stderr, "%s ", currentDateTime().c_str());
            if (Settings.CountColumn >= 0){
                // TODO: Fix floating point misbehaviour on the PER calculation
                fprintf(stderr, "RxRate: %" PRIu64 " MB/s (total: %" PRIu64 " MB) Elapsed %s PktRec = %i ErrCount = %i PktsLost = %i PER %4.3e\n",
                        RxBytes / B1M / interval_s,
                        RxBytesTotal / B1M, elapsed_str,
                        RxPackets, 
                        ErrCount,
                        PktsLost,
                        (1.0 * PktsLost / (RxPackets + PktsLost)));
            } else {
                    if (USecs == 0) {  // protect against divide by zero
                        fprintf(stderr, "Rx rate: %.2f Mbps, rx %" PRIu64 " MB/s (total: %" PRIu64 " MB), Elapsed %s, PktRec = %i\n",
                                RxBytes * 8.0 / B1M,
                                RxBytes / B1M / interval_s,
                                RxBytesTotal / B1M,
                                elapsed_str, 
                                RxPackets);
                    } else {
                        fprintf(stderr, "Rx rate: %.2f Mbps, rx %" PRIu64 " MB/s (total: %" PRIu64 " MB), Elapsed %s, PktRec = %i\n",
                                RxBytes * 8.0 / (USecs / 1000000.0) / B1M,
                                RxBytes / B1M / interval_s,
                                RxBytesTotal / B1M,
                                elapsed_str,
                                RxPackets);
                    }
            }
            RxBytes = 0;
            UpdateTimer.now();
            USecs = UpdateTimer.timeus();
        }
    
  } // for (samples)

    #ifdef _WIN32
    // clean up winsock
      WSACleanup();
    #endif

    return 0;
}
